weave/feedback.py

```diff
@@ -0,0 +1,241 @@
+        creator: The name to display for the originator of the feedback.
+        """
+        if feedback_type.startswith("wandb."):
+            raise ValueError('Feedback type cannot start with "wandb."')
+        feedback = {}
+        feedback.update(payload or {})
+        feedback.update(kwargs)
+        return self._add(feedback_type, feedback, creator)
+
+    def _add(
+        self, feedback_type: str, payload: dict[str, Any], creator: Optional[str]
+    ) -> str:
+        self._maybe_fetch()
+        assert self.items is not None
+        freq = tsi.FeedbackCreateReq(
+            project_id=self.project_id,
+            weave_ref=self.weave_ref,
+            feedback_type=feedback_type,
+            payload=payload,
+            creator=creator,
+        )
+        response = self.client.server.feedback_create(freq)
+
+        # Add to internal items so we don't have to refresh
+        feedback = tsi.Feedback(
+            **freq.dict(),
+            id=response.id,
+            created_at=response.created_at,
+            wb_user_id=response.wb_user_id,
+        )
+        self.items.append(feedback)
+        return response.id
+
+    def add_reaction(self, emoji: str, creator: Optional[str] = None) -> str:
+        return self._add(
+            "wandb.reaction.1",
+            {
+                "emoji": emoji,
+            },
+            creator=creator,
+        )
+
+    def add_note(self, note: str, creator: Optional[str] = None) -> str:
+        return self._add(
+            "wandb.note.1",
+            {
+                "note": note,
+            },
+            creator=creator,
+        )
+
+    # TODO: Consider exposing more flexible query options.
+    def query(self, limit: Optional[int] = None) -> Feedbacks:
+        return Feedbacks(self._query(limit))
+
+    def _query(self, limit: Optional[int] = None) -> list[tsi.Feedback]:
+        query = {
+            "$expr": {
+                "$eq": [
+                    {"$getField": "weave_ref"},
+                    {"$literal": self.weave_ref},
+                ],
+            }
+        }
+        sort_by = [
+            {
+                "field": "created_at",
+                "direction": "asc",
+            },
+        ]
+        req = tsi.FeedbackQueryReq(
+            project_id=self.project_id,
+            query=query,
+            sort_by=sort_by,
+            limit=limit,
+        )
+        response = self.client.server.feedback_query(req)
+        # Response is dicts because API allows user to specify fields, but we don't
+        # expose that in this Python API.
+        return list(tsi.Feedback(**r) for r in response.result)
+
+    def purge(self, feedback_id: str) -> None:
+        req = tsi.FeedbackPurgeReq(
+            project_id=self.project_id,
+            query=Query(
+                **{
+                    "$expr": {
+                        "$eq": [
```

Comment:
might be nice if this also included the ref in the query expression. Else, you could inadvertently purge an id that does not belong to that ref

Reply:
Actually, I'm going to defer this - it's a good idea, but I'd have to first update the server side APIs, because also in the name of safety we put in restrictions on the purge endpoint that the purge query must be only a ref or list of refs.